hpx/parallel/algorithms/unique.hpp

```diff
@@ -0,0 +1,367 @@
+#include <hpx/util/invoke.hpp>
+#include <hpx/util/tagged_pair.hpp>
+
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/is_negative.hpp>
+#include <hpx/parallel/algorithms/detail/predicates.hpp>
+#include <hpx/parallel/algorithms/detail/transfer.hpp>
+#include <hpx/parallel/execution_policy.hpp>
+#include <hpx/parallel/tagspec.hpp>
+#include <hpx/parallel/traits/projected.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/foreach_partitioner.hpp>
+#include <hpx/parallel/util/loop.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+#include <hpx/parallel/util/scan_partitioner.hpp>
+#include <hpx/parallel/util/transfer.hpp>
+#include <hpx/parallel/util/zip_iterator.hpp>
+#include <hpx/util/unused.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <cstring>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <boost/shared_array.hpp>
+
+namespace hpx { namespace parallel { inline namespace v1
+{
+    /////////////////////////////////////////////////////////////////////////////
+    // unique_copy
+    namespace detail
+    {
+        /// \cond NOINTERNAL
+
+        // sequential unique_copy with projection function
+        template <typename FwdIter, typename OutIter, typename Pred, typename Proj>
+        std::pair<FwdIter, OutIter>
+        sequential_unique_copy(FwdIter first, FwdIter last, OutIter dest,
+            Pred && pred, Proj && proj, std::true_type)
+        {
+            if (first == last)
+                return std::make_pair(std::move(last), std::move(dest));
+
+            FwdIter base = first;
+
+            *dest++ = *first;
+
+            while (++first != last)
+            {
+                if (!hpx::util::invoke(pred,
+                    hpx::util::invoke(proj, *base),
+                    hpx::util::invoke(proj, *first)))
+                {
+                    base = first;
+                    *dest++ = *first;
+                }
+            }
+            return std::make_pair(std::move(last), std::move(dest));
+        }
+
+        // sequential unique_copy with projection function
+        template <typename InIter, typename OutIter, typename Pred, typename Proj>
+        std::pair<InIter, OutIter>
+        sequential_unique_copy(InIter first, InIter last, OutIter dest,
+            Pred && pred, Proj && proj, std::false_type)
+        {
+            if (first == last)
+                return std::make_pair(std::move(last), std::move(dest));
+
+            auto base_val = *first;
+
+            *dest++ = *first;
+
+            while (++first != last)
+            {
+                if (!hpx::util::invoke(pred,
+                    hpx::util::invoke(proj, base_val),
+                    hpx::util::invoke(proj, *first)))
+                {
+                    base_val = *first;
+                    *dest++ = *first;
```

Comment:
Same here, calling `*first` more than once without interleaving `++first` is not allowed for input iterators. Not sure if this is worth changing as we know this will not be called with input iterators in the first place...